File: Part8/Section 39 - Artificial Neural Networks (ANN)/a_neural_net.R
#Artificial Neural Network

#Plantilla de Regresion Logística

#--->>Pre Procesado de Datos

#Importar Dataset
df=read.csv('Churn_Modelling.csv')

#filtrado de datasets (filas, columnas)
df=df[,4:14]


#tratamiento de los n/a
# df$Age=ifelse(is.na(df$Age),
#               ave(df$Age,FUN=function(x)mean(x,na.rm = TRUE)),
#               df$Age )
# df$Salary=ifelse(is.na(df$Salary),
#                  ave(df$Salary,FUN=function(x)mean(x,na.rm = TRUE)),
#                  df$Salary )


#codificar variables categóricas a factor y pasar factores a datos numéticos

df$Geography=as.numeric(factor(df$Geography,
                  levels=c("France", "Spain", "Germany"),
                  labels=c(1,2,3)))
df$Gender=as.numeric(factor(df$Gender,
                  levels=c("Female", "Male"),
                  labels=c(0,1)))


#Dividir los datos en entrenamiento y conjunto de test
#install.packages("caTools")
library(caTools)
set.seed(123)
sample=sample.split(df$Exited, SplitRatio = 0.8)
training_set=subset(df, sample==TRUE)
testing_set=subset(df, sample==FALSE)

#Escalado de Valores
training_set[,1:10]=scale(training_set[,1:10])
testing_set[,1:10]=scale(testing_set[,1:10])

#-----<Crear Red Neuronal>-----
library("h2o")
h2o.init(nthreads=-1)
classifier=h2o.deeplearning(y="Exited",
                            training_frame = as.h2o(training_set),
                            activation = "Rectifier",
                            hidden = c(6,6),
                            epochs = 100,
                            train_samples_per_iteration = -2)


#-----<Crear Modelo de Clasificación>----
#crear prediccion si el resultado esperado es una  probabilidad:
prob_pred=h2o.predict(classifier,
                      newdata = as.h2o(testing_set[,-11]))
y_pred=(prob_pred>0.5)
y_pred=as.vector(y_pred)

#Matriz de confucción para evaluar eficiencia 
cm=table(testing_set[,11],
         y_pred)
#Cerrar la sesión de H2o
h2o.shutdown()

File: Part8/Section 39 - Artificial Neural Networks (ANN)/a_neural_net.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu Feb  6 15:25:17 2020

@author: root
"""

#Aritificial Neural Network

#------<PREPROCESADO DE DATOS>----------

#Como importar una libreria

import  numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#importar dataset
df=pd.read_csv("Churn_Modelling.csv")

X=df.iloc[:, 3:13].values
y=df.iloc[:,-1].values

#tratamiento de los nan
"""from sklearn.impute import SimpleImputer as Imputer
imputer=Imputer(missing_values= np.nan, strategy="mean")
imputer=imputer.fit(x[:,1:3])
x[:,1:3]=imputer.transform(x[:,1:3])"""

#datos categóricos ordinales codificar 
"""from sklearn import preprocessing as pre
le_y=pre.LabelEncoder()
y=le_y.fit_transform(y)"""

#datos  categóricos no ordinales

from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.compose import ColumnTransformer
labelencoder_X_1 = LabelEncoder()
X[:, 1] = labelencoder_X_1.fit_transform(X[:, 1])
labelencoder_X_2 = LabelEncoder()
X[:, 2] = labelencoder_X_2.fit_transform(X[:, 2])
onehotencoder = ColumnTransformer([("Geography", OneHotEncoder(), [1])], # The column numbers to be transformed (here is [1])
                                  remainder="passthrough") # Leave the rest of the columns untouched
X = onehotencoder.fit_transform(X)
X = X[:, 1:]
#dividir dataset en conjunto de entrenamiento y testing


from sklearn.model_selection import train_test_split as splitter
x_train, x_test, y_train,  y_test = splitter(X, y, test_size=0.2, random_state=0)

#Escalado de variables
from sklearn.preprocessing import StandardScaler as scaler
scala_x=scaler()
x_train = scala_x.fit_transform(x_train)
x_test = scala_x.transform(x_test)

#------<AJUSTAR MODELOS DE CLASIFICACION>----------
"""
#Ajustar regresion con el el conjunto de entrenamiento

#Crear modelo de clasificación aqui 

from sklearn.linear_model import LogisticRegression
classifier=LogisticRegression(random_state=0)
classifier.fit(x_train,y_train)
print(classifier)


#predicción de nuestros modelos con conjunto de test

y_pred=classifier.predict(x_test)

#predicción de nuestros modelos con nuevo valor
#y_pred=clasifier.predict(6.5)

#crear matriz de confusión (evaluar prediccion)
from sklearn.metrics import confusion_matrix
cm=confusion_matrix(y_test,y_pred)
print(cm)
"""